core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -118,8 +127,27 @@
 
   private final class ContextCancellationListener implements CancellationListener {
     @Override
-    public void cancelled(Context context) {
-      stream.cancel(statusFromCancelled(context));
+    public void cancelled(final Context context) {
+      if (streamCancelledByDeadline) {
+        return;
+      }
+      if (context.getDeadline() == null || !context.getDeadline().isExpired()) {
+        stream.cancel(statusFromCancelled(context));
+      } else {
+        if (deadlineCancellationSendToServerFuture != null) {
```

Comment:
I think we can just leave the `deadlineCancellationSendToServerFuture` and let it do the job.